Well that didn't change much

```
19:49:27 [WARNING] Exception encountered when loading plugin: AlwaysOnline
java.lang.NoClassDefFoundError: net/md_5/bungee/config/YamlConfiguration
    at me.johnnywoof.AlwaysOnline.onEnable(AlwaysOnline.java:37)
    at net.md_5.bungee.api.plugin.PluginManager.loadAndEnablePlugins(PluginManager.java:194)
    at net.md_5.bungee.BungeeCord.start(BungeeCord.java:177)
    at net.md_5.bungee.Bootstrap.main(Bootstrap.java:68)
Caused by: java.lang.ClassNotFoundException: net.md_5.bungee.config.YamlConfiguration
    at net.md_5.bungee.api.plugin.PluginClassloader.loadClass0(PluginClassloader.java:53)
    at net.md_5.bungee.api.plugin.PluginClassloader.loadClass(PluginClassloader.java:27)
    at java.lang.ClassLoader.loadClass(Unknown Source)
    ... 4 more
```

Your change actually caused the plugin to break completely now..

--- a/api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java
+++ b/api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java
@@ -37,4 +37,18 @@ public interface PendingConnection extends Connection
      * @return the accepting listener
      */
     ListenerInfo getListener();
+    
+    /**
+     * Get this connection's online mode.
+     *
+     * @return the online mode
+     */
+    boolean isOnlineMode();
+    
+    /**
+     * Set this connection's online mode.
+     * 
+     * @param onlineMode 
+     */
+    void setOnlineMode(boolean onlineMode);
 }

--- a/api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java
@@ -14,7 +14,8 @@ import net.md_5.bungee.api.plugin.Cancellable;
 @Data
 @ToString(callSuper = false)
 @EqualsAndHashCode(callSuper = false)
-public class PreLoginEvent extends AsyncEvent<PreLoginEvent> implements Cancellable {
+public class PreLoginEvent extends AsyncEvent<PreLoginEvent> implements Cancellable
+{
 
     /**
      * Cancelled state.

--- a/api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/PreLoginEvent.java
@@ -1,0 +1,37 @@
+package net.md_5.bungee.api.event;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import net.md_5.bungee.api.Callback;
+import net.md_5.bungee.api.connection.PendingConnection;
+import net.md_5.bungee.api.plugin.Cancellable;
+
+/**
+ * Event called to represent a player first making their presence and username
+ * known.
+ */
+@Data
+@ToString(callSuper = false)
+@EqualsAndHashCode(callSuper = false)
+public class PreLoginEvent extends AsyncEvent<PreLoginEvent> implements Cancellable {
+
+    /**
+     * Cancelled state.
+     */
+    private boolean cancelled;
+    /**
+     * Message to use when kicking if this event is canceled.
+     */
+    private String cancelReason;
+    /**
+     * Connection attempting to login.
+     */
+    private final PendingConnection connection;
+
+    public PreLoginEvent(PendingConnection connection, Callback<PreLoginEvent> done)
+    {
+        super( done );
+        this.connection = connection;
+    }
+}

--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -4,8 +4,17 @@ import com.google.common.base.Preconditions;
 import io.netty.util.concurrent.ScheduledFuture;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
-import net.md_5.bungee.*;
-import net.md_5.bungee.api.*;
+import net.md_5.bungee.BungeeCord;
+import net.md_5.bungee.EncryptionUtil;
+import net.md_5.bungee.PacketConstants;
+import net.md_5.bungee.UserConnection;
+import net.md_5.bungee.Util;
+import net.md_5.bungee.api.AbstractReconnectHandler;
+import net.md_5.bungee.api.Callback;
+import net.md_5.bungee.api.ChatColor;
+import net.md_5.bungee.api.NewServerPing;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.ServerPing;
 import net.md_5.bungee.api.config.ListenerInfo;
 import net.md_5.bungee.api.config.ServerInfo;
 import net.md_5.bungee.api.connection.PendingConnection;
@@ -13,16 +22,37 @@ import net.md_5.bungee.api.connection.ProxiedPlayer;
 import net.md_5.bungee.api.event.LoginEvent;
 import net.md_5.bungee.api.event.PlayerHandshakeEvent;
 import net.md_5.bungee.api.event.PostLoginEvent;
+import net.md_5.bungee.api.event.PreLoginEvent;
 import net.md_5.bungee.api.event.ProxyPingEvent;
 import net.md_5.bungee.http.HttpClient;
-import net.md_5.bungee.netty.*;
+import net.md_5.bungee.netty.ChannelWrapper;
+import net.md_5.bungee.netty.HandlerBoss;
+import net.md_5.bungee.netty.PacketHandler;
+import net.md_5.bungee.netty.PacketMapping;
+import net.md_5.bungee.netty.PipelineUtils;
 import net.md_5.bungee.netty.decoders.CipherDecoder;
 import net.md_5.bungee.netty.decoders.PacketDecoder;
 import net.md_5.bungee.netty.encoders.CipherEncoder;
 import net.md_5.bungee.protocol.Forge;
 import net.md_5.bungee.protocol.MinecraftInput;
-import net.md_5.bungee.protocol.packet.*;
-import net.md_5.bungee.protocol.packet.protocolhack.*;
+import net.md_5.bungee.protocol.packet.DefinedPacket;
+import net.md_5.bungee.protocol.packet.Packet1Login;
+import net.md_5.bungee.protocol.packet.Packet2Handshake;
+import net.md_5.bungee.protocol.packet.PacketCDClientStatus;
+import net.md_5.bungee.protocol.packet.PacketFAPluginMessage;
+import net.md_5.bungee.protocol.packet.PacketFCEncryptionResponse;
+import net.md_5.bungee.protocol.packet.PacketFDEncryptionRequest;
+import net.md_5.bungee.protocol.packet.PacketFEPing;
+import net.md_5.bungee.protocol.packet.PacketFFKick;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketEncryptionRequest;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketEncryptionResponse;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketHandshake;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketKick;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketLoginStart;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketLoginSuccess;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketPing;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketPingRequest;
+import net.md_5.bungee.protocol.packet.protocolhack.PacketPingResponse;
 
 import javax.crypto.Cipher;
 import javax.crypto.SecretKey;
@@ -290,14 +320,31 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             return;
         }
 
-        if ( this.onlineMode )
+        Callback<PreLoginEvent> callback = new Callback<PreLoginEvent>()
         {
-            unsafe().sendPacket(request172 = EncryptionUtil.encryptRequest172(this.onlineMode)); // packet mapping should take care of this
-            thisState = State.ENCRYPT;
-        } else
-        {
-            finish( true );
-        }
+
+            @Override
+            public void done(PreLoginEvent result, Throwable error) {
+                if(result.isCancelled())
+                {
+                    disconnect17( result.getCancelReason() );
+                }
+                if(ch.isClosed())
+                {
+                    return;
+                }
+                if ( onlineMode )
+                {
+                    unsafe().sendPacket(request172 = EncryptionUtil.encryptRequest172(onlineMode)); // packet mapping should take care of this
+                    thisState = State.ENCRYPT;
+                } else
+                {
+                    finish( true );
+                }
+            }
+        };
+
+        bungee.getPluginManager().callEvent(new PreLoginEvent(InitialHandler.this, callback));
     }
 
     @Override
@@ -343,8 +390,27 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         unsafe().sendPacket( PacketConstants.I_AM_BUNGEE );
         unsafe().sendPacket( PacketConstants.FORGE_MOD_REQUEST );
 
-        unsafe().sendPacket( request164 = EncryptionUtil.encryptRequest164( this.onlineMode ) );
-        thisState = State.ENCRYPT;
+        Callback<PreLoginEvent> callback = new Callback<PreLoginEvent>()
+        {
+
+            @Override
+            public void done(PreLoginEvent result, Throwable error) {
+                if(result.isCancelled())
+                {
+                    disconnect( result.getCancelReason() );
+                }
+                if(ch.isClosed())
+                {
+                    return;
+                }
+
+                unsafe().sendPacket( request164 = EncryptionUtil.encryptRequest164( onlineMode ) );
+
+                thisState = State.ENCRYPT;
+            }
+        };
+
+        bungee.getPluginManager().callEvent( new PreLoginEvent ( InitialHandler.this, callback ) );
     }
 
     @Override

